cache/store.go

```diff
@@ -53,80 +58,108 @@ func (a *kvItem) Less(b btree.Item) bool {
 	return a.key < b.(*kvItem).key
 }
 
-func (s *store) Get(startKey, endKey []byte) ([]*mvccpb.KeyValue, int64, error) {
+func (s *store) Get(startKey, endKey []byte, rev int64) ([]*mvccpb.KeyValue, int64, error) {
+	snapshot, latestRev, err := s.getSnapshot(rev)
+	if err != nil {
+		return nil, 0, err
+	}
+	return snapshot.Range(startKey, endKey), latestRev, nil
+}
+
+func (s *store) getSnapshot(rev int64) (*snapshot, int64, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
 
-	if s.latestRev == 0 {
+	if s.latest.rev == 0 {
 		return nil, 0, ErrNotReady
 	}
+	if rev < 0 {
+		return nil, 0, fmt.Errorf("invalid revision: %d", rev)
+	}
 
-	var out []*mvccpb.KeyValue
-	switch {
-	case len(endKey) == 0:
-		if item := s.tree.Get(probeItemFromBytes(startKey)); item != nil {
-			out = append(out, item.(*kvItem).kv)
-		}
+	snapshot, err := s.findSnapshotForRevLocked(rev)
+	if err != nil {
+		return nil, 0, err
+	}
 
-	case isPrefixScan(endKey):
-		s.tree.AscendGreaterOrEqual(probeItemFromBytes(startKey), func(item btree.Item) bool {
-			out = append(out, item.(*kvItem).kv)
-			return true
-		})
-
-	default:
-		s.tree.AscendRange(probeItemFromBytes(startKey), probeItemFromBytes(endKey), func(item btree.Item) bool {
-			out = append(out, item.(*kvItem).kv)
-			return true
-		})
+	return snapshot, s.latest.rev, nil
+}
+
+func (s *store) findSnapshotForRevLocked(rev int64) (*snapshot, error) {
+	if rev == 0 {
+		rev = s.latest.rev
+	}
+	if rev > s.latest.rev {
+		return nil, rpctypes.ErrFutureRev
 	}
-	return out, s.latestRev, nil
+	oldestRev := s.history.PeekOldest()
+	if rev < oldestRev {
+		return nil, rpctypes.ErrCompacted
+	}
+
+	var targetSnapshot *snapshot
+	s.history.AscendGreaterOrEqual(rev, func(rev int64, snap *snapshot) bool {
+		targetSnapshot = snap
+		return false
+	})
+
+	return targetSnapshot, nil
 }
 
+// Restore replaces state with the bootstrap snapshot and resets history.
 func (s *store) Restore(kvs []*mvccpb.KeyValue, rev int64) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.tree = btree.New(s.degree)
+	s.latest.tree = btree.New(s.degree)
 	for _, kv := range kvs {
-		s.tree.ReplaceOrInsert(newKVItem(kv))
+		s.latest.tree.ReplaceOrInsert(newKVItem(kv))
 	}
-	s.latestRev = rev
+	s.history.RebaseHistory()
+	s.latest.rev = rev
+	s.history.Append(newClonedSnapshot(rev, s.latest.tree))
 }
 
 func (s *store) Apply(events []*clientv3.Event) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	if err := validateRevisions(events, s.latestRev); err != nil {
+	if err := validateRevisions(events, s.latest.rev); err != nil {
 		return err
 	}
 
+	curRev := int64(0)
+	workingTree := s.latest.tree
 	for _, ev := range events {
+		rev := ev.Kv.ModRevision
+		if curRev != 0 && rev != curRev {
+			s.history.Append(newClonedSnapshot(curRev, workingTree))
```

Comment:
Latest should be already be in history so it, clone should be done after applying event.